File: prayer_app_flutter/lib/src/screens/quran_reader_screen.dart
import 'dart:async';
import 'package:flutter/material.dart';
import 'package:material_design_icons_flutter/material_design_icons_flutter.dart';
import '../models/quran_models.dart';
import '../services/quran_api_service.dart';
import '../services/quran_storage_service.dart';
import '../theme/app_theme.dart';
import '../providers/theme_provider.dart';

class QuranReaderScreen extends StatefulWidget {
  final SurahMeta surah;
  final int initialAyahNumber;

  const QuranReaderScreen({
    super.key,
    required this.surah,
    this.initialAyahNumber = 1,
  });

  @override
  State<QuranReaderScreen> createState() => _QuranReaderScreenState();
}

class _QuranReaderScreenState extends State<QuranReaderScreen> {
  final QuranApiService _api = QuranApiService();
  final QuranStorageService _storage = QuranStorageService();
  final ScrollController _scroll = ScrollController();

  List<Ayah> _arabic = [];
  List<Ayah>? _english;
  List<Ayah> _merged = [];
  List<QuranPointer> _bookmarks = [];

  bool _loading = true;
  bool _showTranslation = false;
  bool _translationLoading = false;
  bool _offlineCached = false;
  String? _error;
  double _fontScale = 1.0;

  Timer? _scrollDebounce;
  Timer? _flashTimer;
  int _currentAyah = 1;
  int? _flashAyah;

  @override
  void initState() {
    super.initState();
    _currentAyah = widget.initialAyahNumber;
    _scroll.addListener(_onScroll);
    _loadBookmarks();
    _loadArabic();
  }

  @override
  void dispose() {
    _scrollDebounce?.cancel();
    _flashTimer?.cancel();
    _scroll.dispose();
    super.dispose();
  }

  Future<void> _loadBookmarks() async {
    final bookmarks = await _storage.loadBookmarks();
    if (!mounted) return;
    setState(() => _bookmarks = bookmarks);
  }

  Future<void> _loadArabic() async {
    setState(() {
      _loading = true;
      _error = null;
      _offlineCached = false;
    });

    final cached = await _api.loadCachedArabic(widget.surah.number);
    if (cached != null && cached.isNotEmpty && mounted) {
      setState(() {
        _arabic = cached;
        _merged = cached;
        _loading = false;
      });
      _jumpToAyah(widget.initialAyahNumber);
    }

    try {
      final fresh = await _api.fetchSurahArabic(widget.surah.number);
      if (!mounted) return;
      setState(() {
        _arabic = fresh;
        _merged = _showTranslation && _english != null
            ? _api.mergeArabicAndEnglish(arabic: fresh, english: _english!)
            : fresh;
        _loading = false;
        _offlineCached = false;
      });
      _jumpToAyah(widget.initialAyahNumber);
      if (_showTranslation) {
        _loadTranslation();
      }
    } catch (e) {
      if (!mounted) return;
      if (cached != null && cached.isNotEmpty) {
        setState(() {
          _offlineCached = true;
          _loading = false;
        });
      } else {
        setState(() {
          _error = 'Could not load surah text. Check internet and retry.';
          _loading = false;
        });
      }
      debugPrint('[QuranReader] Arabic fetch failed: $e');
    }
  }

  Future<void> _loadTranslation() async {
    if (_translationLoading) return;

    setState(() => _translationLoading = true);
    final cached = await _api.loadCachedEnglish(widget.surah.number);
    if (cached != null && cached.isNotEmpty && mounted) {
      setState(() {
        _english = cached;
        _merged = _api.mergeArabicAndEnglish(arabic: _arabic, english: cached);
        _translationLoading = false;
      });
    }

    try {
      final fresh = await _api.fetchSurahTranslation(widget.surah.number);
      if (!mounted) return;
      setState(() {
        _english = fresh;
        _merged = _api.mergeArabicAndEnglish(arabic: _arabic, english: fresh);
        _translationLoading = false;
        _offlineCached = false;
      });
    } catch (e) {
      if (!mounted) return;
      setState(() {
        _translationLoading = false;
        if (cached != null && cached.isNotEmpty) {
          _offlineCached = true;
        }
      });
      debugPrint('[QuranReader] Translation fetch failed: $e');
    }
  }

  void _jumpToAyah(int ayahNumber) {
    _flashTimer?.cancel();
    setState(() => _flashAyah = ayahNumber);
    _flashTimer = Timer(const Duration(milliseconds: 1400), () {
      if (mounted) setState(() => _flashAyah = null);
    });
    WidgetsBinding.instance.addPostFrameCallback((_) {
      if (!mounted || !_scroll.hasClients || _merged.isEmpty) return;
      final target = ((ayahNumber - 1).clamp(0, _merged.length - 1)).toDouble();
      _scroll.animateTo(
        target * 150,
        duration: const Duration(milliseconds: 300),
        curve: Curves.easeOut,
      );
    });
  }

  void _onScroll() {
    _scrollDebounce?.cancel();
    _scrollDebounce = Timer(const Duration(milliseconds: 350), () {
      if (_merged.isEmpty) return;
      final idx = (_scroll.offset / 150).round().clamp(0, _merged.length - 1);
      final ayah = _merged[idx].numberInSurah;
      _updateLastRead(ayah);
    });
  }

  Future<void> _updateLastRead(int ayahNumber) async {
    if (_currentAyah == ayahNumber) return;
    _currentAyah = ayahNumber;
    final pointer = QuranPointer(
      surahNumber: widget.surah.number,
      ayahNumber: ayahNumber,
    );
    await _storage.setLastRead(pointer);
    await _storage.pushRecent(pointer);
    if (mounted) setState(() {});
  }

  Future<void> _toggleBookmark() async {
    final pointer = QuranPointer(
      surahNumber: widget.surah.number,
      ayahNumber: _currentAyah,
    );
    final updated = await _storage.toggleBookmark(pointer);
    if (!mounted) return;
    setState(() => _bookmarks = updated);
  }

  bool get _currentBookmarked {
    return _storage.isBookmarked(
      _bookmarks,
      QuranPointer(surahNumber: widget.surah.number, ayahNumber: _currentAyah),
    );
  }

  @override
  Widget build(BuildContext context) {
    final tc = ThemeScope.of(context).current;

    return Scaffold(
      body: Container(
        decoration: BoxDecoration(gradient: appBackgroundGradient(tc)),
        child: SafeArea(
          bottom: false,
          child: Column(
            children: [
              _TopBar(
                surah: widget.surah,
                currentAyah: _currentAyah,
                showTranslation: _showTranslation,
                translationLoading: _translationLoading,
                fontScale: _fontScale,
                isBookmarked: _currentBookmarked,
                onBack: () => Navigator.of(context).pop(),
                onFontDown: () => setState(
                  () => _fontScale = (_fontScale - 0.1).clamp(0.8, 1.6),
                ),
                onFontUp: () => setState(
                  () => _fontScale = (_fontScale + 0.1).clamp(0.8, 1.6),
                ),
                onToggleTranslation: () {
                  setState(() => _showTranslation = !_showTranslation);
                  if (!_showTranslation) {
                    return;
                  }
                  if (_english == null) _loadTranslation();
                },
                onToggleBookmark: _toggleBookmark,
              ),
              if (_offlineCached)
                Container(
                  margin: const EdgeInsets.fromLTRB(20, 8, 20, 8),
                  padding: const EdgeInsets.symmetric(
                    horizontal: 12,
                    vertical: 8,
                  ),
                  decoration: BoxDecoration(
                    color: tc.card,
                    borderRadius: BorderRadius.circular(10),
                    border: Border.all(color: tc.cardBorder),
                  ),
                  child: Row(
                    children: [
                      Icon(MdiIcons.wifiOff, size: 16, color: tc.textMuted),
                      const SizedBox(width: 8),
                      Text(
                        'Offline (cached)',
                        style: AppTypography.caption(tc),
                      ),
                    ],
                  ),
                ),
              Expanded(
                child: _loading && _merged.isEmpty
                    ? Center(child: CircularProgressIndicator(color: tc.accent))
                    : _error != null && _merged.isEmpty
                    ? Center(
                        child: Padding(
                          padding: const EdgeInsets.symmetric(horizontal: 24),
                          child: Column(
                            mainAxisSize: MainAxisSize.min,
                            children: [
                              Text(
                                _error!,
                                textAlign: TextAlign.center,
                                style: AppTypography.caption(tc),
                              ),
                              const SizedBox(height: 12),
                              OutlinedButton(
                                onPressed: _loadArabic,
                                child: const Text('Retry'),
                              ),
                            ],
                          ),
                        ),
                      )
                    : ListView.builder(
                        controller: _scroll,
                        itemCount: _merged.length,
                        padding: const EdgeInsets.fromLTRB(20, 8, 20, 20),
                        itemBuilder: (context, index) {
                          final ayah = _merged[index];
                          return GestureDetector(
                            onTap: () => _updateLastRead(ayah.numberInSurah),
                            child: Container(
                              margin: const EdgeInsets.only(
                                bottom: QuranLayout.ayahItemGap,
                              ),
                              padding: const EdgeInsets.all(
                                QuranLayout.ayahItemPadding,
                              ),
                              decoration: BoxDecoration(
                                color: tc.card,
                                borderRadius: BorderRadius.circular(
                                  QuranLayout.cardRadius,
                                ),
                                border: Border.all(
                                  color: _flashAyah == ayah.numberInSurah
                                      ? tc.accent.withValues(alpha: 0.9)
                                      : ayah.numberInSurah == _currentAyah
                                      ? tc.accent.withValues(alpha: 0.5)
                                      : tc.cardBorder,
                                ),
                              ),
                              child: Column(
                                crossAxisAlignment: CrossAxisAlignment.stretch,
                                children: [
                                  Text(
                                    '${ayah.numberInSurah}',
                                    style: AppTypography.caption(tc).copyWith(
                                      color: tc.accent,
                                      fontWeight: FontWeight.w600,
                                    ),
                                  ),
                                  const SizedBox(height: 8),
                                  Text(
                                    ayah.textAr,
                                    textAlign: TextAlign.right,
                                    textDirection: TextDirection.rtl,
                                    style: TextStyle(
                                      fontFamily: 'serif',
                                      fontSize:
                                          QuranLayout.ayahArabicSize *
                                          _fontScale,
                                      height: 1.9,
                                      color: tc.textPrimary,
                                    ),
                                  ),
                                  if (_showTranslation &&
                                      (ayah.textEn?.isNotEmpty ?? false)) ...[
                                    const SizedBox(height: 12),
                                    Text(
                                      ayah.textEn!,
                                      style: AppTypography.body(tc).copyWith(
                                        fontSize:
                                            QuranLayout.ayahTranslationSize *
                                            _fontScale,
                                        color: tc.textMuted,
                                        height: 1.45,
                                      ),
                                    ),
                                  ],
                                ],
                              ),
                            ),
                          );
                        },
                      ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

class _TopBar extends StatelessWidget {
  final SurahMeta surah;
  final int currentAyah;
  final bool showTranslation;
  final bool translationLoading;
  final bool isBookmarked;
  final double fontScale;
  final VoidCallback onBack;
  final VoidCallback onFontDown;
  final VoidCallback onFontUp;
  final VoidCallback onToggleTranslation;
  final VoidCallback onToggleBookmark;

  const _TopBar({
    required this.surah,
    required this.currentAyah,
    required this.showTranslation,
    required this.translationLoading,
    required this.isBookmarked,
    required this.fontScale,
    required this.onBack,
    required this.onFontDown,
    required this.onFontUp,
    required this.onToggleTranslation,
    required this.onToggleBookmark,
  });

  @override
  Widget build(BuildContext context) {
    final tc = ThemeScope.of(context).current;
    return Padding(
      padding: const EdgeInsets.fromLTRB(12, 8, 12, 8),
      child: Row(
        children: [
          IconButton(
            onPressed: onBack,
            icon: Icon(MdiIcons.arrowLeft, color: tc.textPrimary),
          ),
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(
                  surah.englishName,
                  maxLines: 1,
                  overflow: TextOverflow.ellipsis,
                  style: AppTypography.body(
                    tc,
                  ).copyWith(fontWeight: FontWeight.w600),
                ),
                Text(
                  '${surah.nameAr} · Ayah $currentAyah',
                  maxLines: 1,
                  overflow: TextOverflow.ellipsis,
                  style: AppTypography.caption(tc),
                ),
              ],
            ),
          ),
          IconButton(
            onPressed: onFontDown,
            icon: Icon(MdiIcons.formatFontSizeDecrease, color: tc.textPrimary),
          ),
          IconButton(
            onPressed: onFontUp,
            icon: Icon(MdiIcons.formatFontSizeIncrease, color: tc.textPrimary),
          ),
          IconButton(
            onPressed: onToggleTranslation,
            icon: translationLoading
                ? SizedBox(
                    width: 18,
                    height: 18,
                    child: CircularProgressIndicator(
                      strokeWidth: 2,
                      color: tc.accent,
                    ),
                  )
                : Icon(
                    showTranslation
                        ? MdiIcons.translate
                        : MdiIcons.translateOff,
                    color: showTranslation ? tc.accent : tc.textPrimary,
                  ),
          ),
          IconButton(
            onPressed: onToggleBookmark,
            icon: Icon(
              isBookmarked ? MdiIcons.bookmark : MdiIcons.bookmarkOutline,
              color: isBookmarked ? tc.accent : tc.textPrimary,
            ),
          ),
          const SizedBox(width: 2),
          Text(
            '${fontScale.toStringAsFixed(1)}x',
            style: AppTypography.caption(tc),
          ),
        ],
      ),
    );
  }
}

File: prayer_app_expo/src/screens/QuranReaderScreen.js
import React, { useEffect, useMemo, useRef, useState } from 'react';
import {
    View,
    Text,
    StyleSheet,
    TouchableOpacity,
    FlatList,
    ActivityIndicator,
} from 'react-native';
import { MaterialCommunityIcons } from '@expo/vector-icons';
import { useTheme } from '../providers/ThemeProvider';
import { QuranLayout, interFont, getTypography } from '../theme/theme';
import {
    fetchSurahArabic,
    fetchSurahTranslation,
    loadCachedArabic,
    loadCachedEnglish,
    mergeArabicAndEnglish,
} from '../services/quranApi';
import {
    loadBookmarks,
    setLastRead,
    pushRecent,
    toggleBookmark,
    isBookmarked,
} from '../services/quranStorageService';

const AVG_AYAH_HEIGHT = 150;

export default function QuranReaderScreen({ surah, initialAyahNumber = 1, onBack }) {
    const { theme: tc } = useTheme();
    const typo = getTypography(tc);
    const listRef = useRef(null);

    const [arabic, setArabic] = useState([]);
    const [english, setEnglish] = useState(null);
    const [bookmarks, setBookmarks] = useState([]);

    const [showTranslation, setShowTranslation] = useState(false);
    const [translationLoading, setTranslationLoading] = useState(false);
    const [fontScale, setFontScale] = useState(1.0);

    const [loading, setLoading] = useState(true);
    const [offlineCached, setOfflineCached] = useState(false);
    const [error, setError] = useState(null);
    const [currentAyah, setCurrentAyah] = useState(initialAyahNumber);
    const [flashAyah, setFlashAyah] = useState(initialAyahNumber);
    const flashTimerRef = useRef(null);

    useEffect(() => {
        (async () => {
            const saved = await loadBookmarks();
            setBookmarks(saved);
        })();
    }, []);

    useEffect(() => {
        loadArabic();
    }, [surah.number]);

    const merged = useMemo(() => {
        if (showTranslation && english) {
            return mergeArabicAndEnglish(arabic, english);
        }
        return arabic;
    }, [arabic, english, showTranslation]);

    useEffect(() => {
        if (merged.length > 0) {
            jumpToAyah(initialAyahNumber);
        }
    }, [merged.length]);

    useEffect(() => {
        return () => {
            if (flashTimerRef.current) clearTimeout(flashTimerRef.current);
        };
    }, []);

    async function loadArabic() {
        setLoading(true);
        setOfflineCached(false);
        setError(null);

        const cached = await loadCachedArabic(surah.number);
        if (cached?.length) {
            setArabic(cached);
            setLoading(false);
        }

        try {
            const fresh = await fetchSurahArabic(surah.number);
            setArabic(fresh);
            setLoading(false);
            setOfflineCached(false);
            if (showTranslation) loadTranslation();
        } catch (e) {
            if (cached?.length) {
                setOfflineCached(true);
                setLoading(false);
            } else {
                setError('Failed to load surah text');
                setLoading(false);
            }
        }
    }

    async function loadTranslation() {
        if (translationLoading) return;
        setTranslationLoading(true);

        const cached = await loadCachedEnglish(surah.number);
        if (cached?.length) {
            setEnglish(cached);
            setTranslationLoading(false);
        }

        try {
            const fresh = await fetchSurahTranslation(surah.number);
            setEnglish(fresh);
            setTranslationLoading(false);
            setOfflineCached(false);
        } catch (e) {
            setTranslationLoading(false);
            if (cached?.length) setOfflineCached(true);
        }
    }

    function jumpToAyah(ayahNumber) {
        if (flashTimerRef.current) clearTimeout(flashTimerRef.current);
        setFlashAyah(ayahNumber || 1);
        flashTimerRef.current = setTimeout(() => setFlashAyah(null), 1400);
        const idx = Math.max(0, Math.min((ayahNumber || 1) - 1, Math.max(0, merged.length - 1)));
        setTimeout(() => {
            try {
                listRef.current?.scrollToIndex({ index: idx, animated: true, viewPosition: 0.05 });
            } catch (_) {
                // best effort
            }
        }, 60);
    }

    async function updateLastRead(ayahNumber) {
        if (!ayahNumber || ayahNumber === currentAyah) return;
        setCurrentAyah(ayahNumber);
        const pointer = { surahNumber: surah.number, ayahNumber };
        await setLastRead(pointer);
        await pushRecent(pointer);
    }

    async function onToggleBookmark() {
        const pointer = { surahNumber: surah.number, ayahNumber: currentAyah };
        const next = await toggleBookmark(pointer);
        setBookmarks(next);
    }

    function updateFromOffset(y) {
        if (!merged.length) return;
        const idx = Math.max(0, Math.min(Math.round(y / AVG_AYAH_HEIGHT), merged.length - 1));
        updateLastRead(merged[idx].numberInSurah);
    }

    const currentBookmarked = isBookmarked(bookmarks, {
        surahNumber: surah.number,
        ayahNumber: currentAyah,
    });

    return (
        <View style={{ flex: 1 }}>
            <View style={styles.topBar}>
                <TouchableOpacity onPress={onBack} hitSlop={10}>
                    <MaterialCommunityIcons name="arrow-left" size={24} color={tc.textPrimary} />
                </TouchableOpacity>
                <View style={{ flex: 1, marginHorizontal: 10 }}>
                    <Text style={[typo.body, { fontFamily: interFont('600') }]} numberOfLines={1}>{surah.englishName}</Text>
                    <Text style={[typo.caption]} numberOfLines={1}>{surah.nameAr} · Ayah {currentAyah}</Text>
                </View>
                <TouchableOpacity onPress={() => setFontScale((s) => Math.max(0.8, +(s - 0.1).toFixed(1)))}>
                    <MaterialCommunityIcons name="format-font-size-decrease" size={QuranLayout.topActionIconSize} color={tc.textPrimary} />
                </TouchableOpacity>
                <View style={{ width: 8 }} />
                <TouchableOpacity onPress={() => setFontScale((s) => Math.min(1.6, +(s + 0.1).toFixed(1)))}>
                    <MaterialCommunityIcons name="format-font-size-increase" size={QuranLayout.topActionIconSize} color={tc.textPrimary} />
                </TouchableOpacity>
                <View style={{ width: 8 }} />
                <TouchableOpacity
                    onPress={() => {
                        const next = !showTranslation;
                        setShowTranslation(next);
                        if (next && !english) loadTranslation();
                    }}
                >
                    {translationLoading ? (
                        <ActivityIndicator size="small" color={tc.accent} />
                    ) : (
                        <MaterialCommunityIcons
                            name={showTranslation ? 'translate' : 'translate-off'}
                            size={QuranLayout.topActionIconSize}
                            color={showTranslation ? tc.accent : tc.textPrimary}
                        />
                    )}
                </TouchableOpacity>
                <View style={{ width: 8 }} />
                <TouchableOpacity onPress={onToggleBookmark}>
                    <MaterialCommunityIcons
                        name={currentBookmarked ? 'bookmark' : 'bookmark-outline'}
                        size={QuranLayout.topActionIconSize}
                        color={currentBookmarked ? tc.accent : tc.textPrimary}
                    />
                </TouchableOpacity>
            </View>

            {offlineCached && (
                <View style={[styles.banner, { backgroundColor: tc.card, borderColor: tc.cardBorder }]}> 
                    <MaterialCommunityIcons name="wifi-off" size={16} color={tc.textMuted} />
                    <Text style={[typo.caption, { marginLeft: 8 }]}>Offline (cached)</Text>
                </View>
            )}

            {loading && merged.length === 0 ? (
                <View style={styles.center}><ActivityIndicator size="large" color={tc.accent} /></View>
            ) : error && merged.length === 0 ? (
                <View style={styles.center}>
                    <Text style={[typo.caption, { textAlign: 'center', paddingHorizontal: 24 }]}>{error}</Text>
                    <View style={{ height: 12 }} />
                    <TouchableOpacity style={[styles.retryBtn, { borderColor: tc.cardBorder }]} onPress={loadArabic}>
                        <Text style={[typo.body, { fontSize: 14 }]}>Retry</Text>
                    </TouchableOpacity>
                </View>
            ) : (
                <FlatList
                    ref={listRef}
                    data={merged}
                    keyExtractor={(item) => `ayah-${item.numberInSurah}`}
                    contentContainerStyle={{ paddingHorizontal: 20, paddingBottom: 24, paddingTop: 8 }}
                    onMomentumScrollEnd={(e) => updateFromOffset(e.nativeEvent.contentOffset.y)}
                    onScrollEndDrag={(e) => updateFromOffset(e.nativeEvent.contentOffset.y)}
                    onScrollToIndexFailed={() => {
                        // best effort only
                    }}
                    renderItem={({ item }) => (
                        <TouchableOpacity
                            activeOpacity={0.8}
                            onPress={() => updateLastRead(item.numberInSurah)}
                            style={[
                                styles.ayahItem,
                                {
                                    backgroundColor: tc.card,
                                    borderColor: item.numberInSurah === flashAyah
                                        ? `${tc.accent}EE`
                                        : item.numberInSurah === currentAyah
                                            ? `${tc.accent}88`
                                            : tc.cardBorder,
                                },
                            ]}
                        >
                            <Text style={[typo.caption, { color: tc.accent, fontFamily: interFont('600') }]}>{item.numberInSurah}</Text>
                            <View style={{ height: 8 }} />
                            <Text style={[
                                styles.arabic,
                                {
                                    color: tc.textPrimary,
                                    fontSize: QuranLayout.ayahArabicSize * fontScale,
                                },
                            ]}>
                                {item.textAr}
                            </Text>
                            {showTranslation && item.textEn ? (
                                <>
                                    <View style={{ height: 12 }} />
                                    <Text
                                        style={{
                                            color: tc.textMuted,
                                            fontFamily: interFont('400'),
                                            fontSize: QuranLayout.ayahTranslationSize * fontScale,
                                            lineHeight: QuranLayout.ayahTranslationSize * fontScale * 1.45,
                                        }}
                                    >
                                        {item.textEn}
                                    </Text>
                                </>
                            ) : null}
                        </TouchableOpacity>
                    )}
                />
            )}
        </View>
    );
}

const styles = StyleSheet.create({
    topBar: {
        flexDirection: 'row',
        alignItems: 'center',
        paddingHorizontal: 12,
        paddingVertical: 8,
    },
    banner: {
        marginHorizontal: 20,
        marginTop: 8,
        marginBottom: 4,
        borderWidth: 1,
        borderRadius: 10,
        paddingHorizontal: 12,
        paddingVertical: 8,
        flexDirection: 'row',
        alignItems: 'center',
    },
    center: {
        flex: 1,
        alignItems: 'center',
        justifyContent: 'center',
    },
    retryBtn: {
        height: 34,
        borderRadius: 17,
        borderWidth: 1,
        alignItems: 'center',
        justifyContent: 'center',
        paddingHorizontal: 14,
    },
    ayahItem: {
        marginBottom: QuranLayout.ayahItemGap,
        padding: QuranLayout.ayahItemPadding,
        borderRadius: QuranLayout.cardRadius,
        borderWidth: 1,
    },
    arabic: {
        fontFamily: 'serif',
        textAlign: 'right',
        lineHeight: QuranLayout.ayahArabicSize * 1.9,
    },
});
